### ISSUE TITLE
Yolov11 COCO val score replication.

### ISSUE THREAD

**Issue by bibhabasumohapatra** [ultralytics/ultralytics#20772]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

Hi ultralytics,

I used my own coco_inference script using pycocotools and taking outputs from results[0].boxes.xyxy. I dont know the reason why scores in the document not matching with the scores from pycocotools. mAP 50-95 is 39.5 in docs and in my script it is 0.311.

Please see I did not made any changes to the model or the outputs. Why it is not reproducing the results any reasons?

### Environment

```
OS                  Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.9
Install             git
Path                /home/bibhabasum/projects/IIIT/ultralytics/ultralytics
RAM                 11.61 GB
Disk                110.0/1006.9 GB
CPU                 unknown
CPU count           8
GPU                 NVIDIA GeForce GTX 1650 Ti, 4096MiB
GPU count           1
CUDA                12.4
```

### Minimal Reproducible Example

```
import os
import json
import torch
from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval
from tqdm import tqdm
from PIL import Image
import torchvision.transforms as T

# Replace this with your actual mode
import numpy as np

# Configuration
COCO_PATH = '/home/bibhabasum/projects/IIIT/data/coco/'
IMG_DIR = os.path.join(COCO_PATH, 'val2017')
ANN_FILE = os.path.join(COCO_PATH, 'annotations/instances_val2017.json')
DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'

# Load model
from ultralytics import YOLO
model = YOLO("yolo11n.pt")

# Load COCO
coco = COCO(ANN_FILE)
img_ids = coco.getImgIds()

results = []
n = []
for img_id in tqdm(img_ids):
    img_info = coco.loadImgs(img_id)[0]
    img_path = os.path.join(IMG_DIR, img_info['file_name'])
    image = Image.open(img_path).convert("RGB")
    image = image.resize((640, 640))  # Resize to model input size
    image = np.asarray(image).astype(np.float32) / 255.0  # Normalize to [0, 1]
    image = torch.tensor(image).permute(2, 0, 1).unsqueeze(0).to(DEVICE)

    model.eval()
    with torch.no_grad():
        preds = model(image) # Get predictions in [cx, cy, w, h, conf, cls...]
        detections = preds[0].boxes.xyxy
        scores = preds[0].boxes.conf
        clss = preds[0].boxes.cls
    # Convert to COCO format
    for i,det in enumerate(detections):
        x1, y1, x2, y2 = det
        width = x2 - x1
        height = y2 - y1
        # # Scale bbox back to original image dimensions
        orig_width, orig_height = img_info['width'], img_info['height']
        scale_x = orig_width / 640
        scale_y = orig_height / 640

        score, cls_id = scores[i],clss[i]
        
        # Map class ID to category ID using classes_dict
        from classes_dict import classes_dict
        class_name = classes_dict[int(cls_id)]
        category_id = coco.getCatIds(catNms=[class_name])[0]

        x1 = x1 * scale_x
        y1 = y1 * scale_y
        width = width * scale_x
        height = height * scale_y
        
        results.append({
            "image_id": img_id,
            "category_id": category_id,
            "bbox": [x1.item(), y1.item(), width.item(), height.item()],
            "score": score.item(),
        })
    # print(img_id, results)
    # break
# print(set(n))
# Save results
with open('coco_results.json', 'w') as f:
    json.dump(results, f)

# Evaluate
coco_dt = coco.loadRes('coco_results.json')
coco_eval = COCOeval(coco, coco_dt, iouType='bbox')
coco_eval.evaluate()
coco_eval.accumulate()
coco_eval.summarize()
```

### Additional

```
classes_dict = {
    0: "person",
    1: "bicycle",
    2: "car",
    3: "motorcycle",
    4: "airplane",
    5: "bus",
    6: "train",
    7: "truck",
    8: "boat",
    9: "traffic light",
    10: "fire hydrant",
    11: "stop sign",
    12: "parking meter",
    13: "bench",
    14: "bird",
    15: "cat",
    16: "dog",
    17: "horse",
    18: "sheep",
    19: "cow",
    20: "elephant",
    21: "bear",
    22: "zebra",
    23: "giraffe",
    24: "backpack",
    25: "umbrella",
    26: "handbag",
    27: "tie",
    28: "suitcase",
    29: "frisbee",
    30: "skis",
    31: "snowboard",
    32: "sports ball",
    33: "kite",
    34: "baseball bat",
    35: "baseball glove",
    36: "skateboard",
    37: "surfboard",
    38: "tennis racket",
    39: "bottle",
    40: "wine glass",
    41: "cup",
    42: "fork",
    43: "knife",
    44: "spoon",
    45: "bowl",
    46: "banana",
    47: "apple",
    48: "sandwich",
    49: "orange",
    50: "broccoli",
    51: "carrot",
    52: "hot dog",
    53: "pizza",
    54: "donut",
    55: "cake",
    56: "chair",
    57: "couch",
    58: "potted plant",
    59: "bed",
    60: "dining table",
    61: "toilet",
    62: "tv",
    63: "laptop",
    64: "mouse",
    65: "remote",
    66: "keyboard",
    67: "cell phone",
    68: "microwave",
    69: "oven",
    70: "toaster",
    71: "sink",
    72: "refrigerator",
    73: "book",
    74: "clock",
    75: "vase",
    76: "scissors",
    77: "teddy bear",
    78: "hair drier",
    79: "toothbrush",
}
```

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.311
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.418
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.343
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.126
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.337
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.484
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.257
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.360
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.363
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.139
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @bibhabasumohapatra, thank you for your detailed report and for using Ultralytics 🚀! We appreciate the time you've taken to provide your script and environment details.

If this is a 🐛 Bug Report, please ensure your [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) includes only the essential code required to reproduce the issue. This helps us debug more efficiently.

For replication and evaluation issues, please double-check that you are using the latest official [ultralytics](https://github.com/ultralytics/ultralytics) package and model weights as described in our [Docs](https://docs.ultralytics.com/). Sometimes, differences in preprocessing, postprocessing, or confidence thresholds can affect mAP results.

An Ultralytics engineer will review your issue and assist you further soon 🤖.

Join our community for discussion and troubleshooting:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
- [Discourse](https://community.ultralytics.com/) for in-depth questions
- [Subreddit](https://reddit.com/r/Ultralytics) for sharing and learning

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to make sure your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your report and patience while we look into your issue!

**Comment by Y-T-G** [comment]
Prediction applies a confidence threshold of 0.25. Validation applies a confidence threshold of 0.001.
https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation

You need to use the same threshold. And it would still not be the same because validation provides multi-label output, but prediction only returns best class per object.

**Comment by bibhabasumohapatra** [comment]
@Y-T-G thanks for the quick response, Actually I was replicating my own YOLOv11 and wanted to test the ultralytics model with the same script. I tried by taking the Detect layer output (x,y,w,h,conf,cls) - 
```
layer.stride = torch.tensor([ 8., 16., 32.])
out = layer([p16, p19, x])
out = layer.postprocess(out[0].permute(0, 2, 1))
```

 and applying NMS myself with conf_thres = 0.001 and iou_thres = 0.65 it still gave me 37.2 mAP and not 39.5 mAP?

it would be great if you can suggest me something or change that can reproduce the result from above script.
